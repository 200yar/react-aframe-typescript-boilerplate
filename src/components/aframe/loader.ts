export const loaderComponentConfig =  {

  schema: {'default': ''},

  /**
   * Do not use arrow function here
   * "this: any" is needed for the typescript compiler
   * https://github.com/Microsoft/TypeScript/issues/16016#issuecomment-303462193
   */
  init: function(this: any) {
    this.el.addEventListener('loaded', () => {
      console.log('scene loaded');
      const loader = document.getElementById('loader') as HTMLDivElement;
      (loader.parentElement as HTMLDivElement).removeChild(loader);
    })
  }
}
