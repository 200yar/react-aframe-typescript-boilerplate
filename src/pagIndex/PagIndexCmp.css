.dialog-subtitle {
  font-size: small;
  padding-bottom: 10px;
}
.icon-info {
  vertical-align: middle;
  padding: 5px;
  border-radius: 4px;
  border: 2px solid;
  background-color: grey;
  margin: 3px;
}
fieldset {
  border-radius: 4px;
  margin-bottom: 5px;
}
legend {
  text-align: left;
  font-size: small;
}
.dialog-close-bottom {
  cursor: pointer;
  text-decoration: underline;
  padding: 10px;
}
.dialog-link {
  color: cornflowerblue;
  text-decoration: none;
}
