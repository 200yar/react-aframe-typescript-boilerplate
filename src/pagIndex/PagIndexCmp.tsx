//todo: crear un component de react que sea un link con imagen (parecido al componente portal)
//todo: react dialog
//todo: react menu
//todo: arreglar lo de que da errores al poner atributos no html5. Por ejemplo <img crossorigin="anonymous">
//todo: cambiar color de theme de chrome android en index.html
//todo: añadir click de mouse (raytracer component) a boton play de video 2d control
//todo: usar "ref" en <a-camera> a ver si se arregla el funcionamiento de las animaciones
//todo: gestionar mejor el estado de la entidad camara
//todo: poner attributos que faltan a <video> en forma de props
//todo: camera position as component state
//todo: probar a eliminar completamente tslint del proyecto
//todo: able to choose to run with or without tslint
//todo: mejorar el manejo del estado "orbitControls"
//todo: añadir pagina con controles material design
//todo: usar aframe.js no minificado durante development time
//todo: reducir tamaño de imagenes
//todo: creditos
//todo: usar una imagen de fondo mejor
//todo: notificar al usuario manejo de controles: rueda del raton hace zoom, etc.
//todo: hacer test
//todo: custom event polyfill
//todo: mouse cursor pointer on <a-link>

/// <reference path="../index.d.ts"/>
import React from 'react';
import 'aframe-orbit-controls-component-2/dist/aframe-orbit-controls-component';
import Loader from "../components/loader/LoaderCmp";
import Dialog from "../components/dialog/DialogCmp";
import SideMenu from "../components/sideMenu/SideMenuCmp";


interface IState {
  orbitControls: {
    autoRotate?: boolean,
    target?: string,
    enableDamping?: boolean,
    dampingFactor?: number,
    rotateSpeed?: number,
    minDistance?: number,
    maxDistance?: number,
    rotateTo?: {x: number, y: number, z: number}
  };
}

export default class PagIndexCmp extends React.Component<any, IState> {

  public state = {
    orbitControls: {
      autoRotate: false,
      target: '#entityGroup',
      enableDamping: true,
      dampingFactor: 0.14,
      rotateSpeed: 0.1,
      autoRotateSpeed: 0.25,
      zoomSpeed: 0.5,
      minDistance: 3,
      maxDistance: 100,
    }
  };

  public props: any;

  public refs: {loader: Loader, scene: AFrame.Entity, dialog: Dialog, sideMenu: SideMenu}

  public constructor(props: any) {
    super(props);
    this.onClickLink1 = this.onClickLink1.bind(this);
    this.onClickLink2 = this.onClickLink2.bind(this);
    this.onClickLink3 = this.onClickLink3.bind(this);
    this.openDialog = this.openDialog.bind(this);
    this.closeDialog = this.closeDialog.bind(this);
    this.startAnimation = this.startAnimation.bind(this);
    this.stopAnimation = this.stopAnimation.bind(this);
  }

  public componentDidMount() {

    // AFRAME Events must be defined in componentDidMount().
    // The others React Events are defined in elements as always

    const aHtmlTags = document.querySelectorAll(".rotate-camera") as HTMLCollection;
    Array.from(aHtmlTags).forEach( (aTag: HTMLAnchorElement) => {
      aTag.addEventListener('click', (event: any) => {
        const position = event.target.dataset.position;
        //todo: revisar esto, se producen estados incosistentes
        this.setState( {orbitControls: {rotateTo: position}} )
      });
    });

    const aFrameLinks = document.querySelectorAll('a-link') as HTMLCollection;
    Array.from(aFrameLinks).forEach( (aLink: AFrame.Entity) => {
      aLink.addEventListener('mouseenter', () => {
        aLink.setAttribute('scale', {x: 1.1, y: 1.1, z: 1.1});
      });
      aLink.addEventListener('mouseleave', () => {
        aLink.setAttribute('scale', {x: 1, y: 1, z: 1})
      })
    });
    this.refs.loader.hideWhen(this.refs.scene, 'loaded');
  }

  private objToString(component: Object): string {
    return AFRAME.utils.styleParser.stringify(component);
  }

  private stopAnimation(): void {
    this.setState({orbitControls: {autoRotate: false}});
  }

  private startAnimation(): void {
    this.setState({orbitControls: {autoRotate: true}});
  }

  private onClickLink1(event: Event) {
    event.preventDefault();
    this.refs.loader.show();
    this.props.history.push('/2dvideo');
  }

  private onClickLink2(event: Event) {
    event.preventDefault();
    this.refs.loader.show();
    this.props.history.push('/360video');
  }

  private onClickLink3 (event: Event) {
    event.preventDefault();
    this.refs.loader.show();
    this.props.history.push('/3dmodel');
  }

  private openDialog() {
    this.refs.dialog.show();
  }

  private closeDialog() {
    this.refs.dialog.hide();
  }

  private openSideMenu() {
    this.refs.sideMenu.show();
  }

  private closeSideMenu() {
    this.refs.sideMenu.hide();
  }

  public render() {
    return (
      <div>

        <Loader ref="loader">Loading</Loader>

        {/*<Dialog ref="dialog">*/}
          {/*Dialog*/}
          {/*<br/><br/>*/}
          {/*<div onClick={ this.closeDialog } style={ {cursor: 'pointer', textDecoration: 'underline'} }>*/}
            {/*Close*/}
          {/*</div>*/}
        {/*</Dialog>*/}

        <SideMenu ref="sideMenu">side menu</SideMenu>

        <div className="top-menu">
          <a className="top-menu-item rotate-camera" data-position="0.17 4.14 2.79">Position 1</a>
          <a className="top-menu-item rotate-camera" data-position="3.48 0.57 0.15">Position 2</a>
          <a className="top-menu-item rotate-camera" data-position="-2.89 -2.51 3.20">Position 3</a>
          <a className="top-menu-item" onClick={ this.openSideMenu.bind(this) }>Open Menu</a>
          {/*<a onClick={ this.startAnimation } className="top-menu-item">Start Rotation</a>*/}
          {/*<a onClick={ this.stopAnimation } className="top-menu-item">Stop Rotation</a>*/}
        </div>

        <a-scene id="scene" ref="scene" raycaster="far: 100; objects: [link], [url]; interval: 200" cursor="rayOrigin: mouse">

          <a-assets>
            <img id="sky" src="img/1.jpg"/>
            <img id="link1" src="img/sea.jpg"/>
            <img id="link2" src="img/4.jpg"/>
            <img id="link3" src="img/7.jpg"/>
            <img id="aframeArena" src="img/aframeArena.png"/>
          </a-assets>

          <a-sky src="#sky" rotation="0 -90 0"/>

          <a-entity id="camera" camera="fov: 80; zoom: 1" position="0 -0.2 5"
            orbit-controls={ this.objToString(this.state.orbitControls) }/>

          <a-entity id="entityGroup">
            <a-plane position="0 -1 0" rotation="-90 0 0" width="6" height="6" src="#aframeArena"/>
            <a-link id="link1" image="#link1" onClick={ this.onClickLink1 } href="#" title="2D Video" position="-3 1 0"/>
            <a-link id="link2" image="#link2" onClick={ this.onClickLink2 } href="#" title="360 Video" position="0 1 0"/>
            <a-link id="link3" image="#link3" onClick={ this.onClickLink3 } href="#" title="Model Animation" position="3 1 0"/>
          </a-entity>

        </a-scene>
      </div>
    );
  }
}