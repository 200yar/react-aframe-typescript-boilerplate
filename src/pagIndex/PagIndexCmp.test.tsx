import 'aframe';
import * as React from 'react';
import {PagIndexCmp} from "./PagIndexCmp";
import TestRenderer from 'react-test-renderer';

describe('<PagIndexCmp/>', () => {

  it('Matches snapshot', () => {
    const component = TestRenderer.create(
      <PagIndexCmp>Pag Index Content</PagIndexCmp>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

});

