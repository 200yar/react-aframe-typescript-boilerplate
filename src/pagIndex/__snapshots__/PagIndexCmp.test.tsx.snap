// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PagIndexCmp/> Matches snapshot 1`] = `
<main>
  <div
    className="fade-in"
    id="loader-element"
    style={
      Object {
        "position": "absolute",
        "textAlign": "center",
        "top": "41%",
        "width": "100%",
        "zIndex": 11,
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": "4px",
          "boxShadow": "grey 0 0 10px 0",
          "color": "black",
          "display": "block",
          "fontFamily": "verdana, sans-serif",
          "fontSize": "small",
          "margin": "auto",
          "padding": "20px",
          "textAlign": "center",
          "width": "70px",
        }
      }
    >
      <img
        src="img/history3.svg"
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      />
      <div
        style={
          Object {
            "paddingTop": "5px",
          }
        }
      >
        Loading
      </div>
    </div>
  </div>
  <div
    className="dialogContainer"
  >
    <div
      className="dialogContent scale-up"
    >
      <div
        className="dialog-title"
      />
      <div
        className="dialog-subtitle"
      >
        This is a demostration of ReactJS, AFrame and TypeScript integration. No real functionality is implemented
      </div>
      <div
        className="dialog-subtitle"
      >
        For a full experience, use a
        <a
          className="dialog-link"
          href="https://www.amazon.com/s/ref=nb_sb_noss?url=search-alias%3Daps&field-keywords=vr+viewer"
          target="_blank"
        >
           VR Viewer
        </a>
      </div>
      <fieldset>
        <legend>
          Pan
        </legend>
        <div>
          <img
            className="icon-info"
            src="mouse-move.svg"
          />
           Mouse Pointer
        </div>
        <div>
          <img
            className="icon-info"
            src="move-gesture.svg"
          />
           Hand Gesture
        </div>
      </fieldset>
      <fieldset>
        <legend>
          Zoom
        </legend>
        <div>
          <img
            className="icon-info"
            src="mouse-wheel.svg"
          />
           Mouse Wheel
        </div>
        <div>
          <img
            className="icon-info"
            src="zoom-gesture.svg"
          />
           Hand Gesture
        </div>
      </fieldset>
      <div
        className="dialog-close-bottom"
        onClick={[Function]}
      >
        Close
      </div>
    </div>
  </div>
  <div
    className="sideMenuContainer slide-in-left"
  >
    <div
      className="sideMenuContent"
    >
      <h1
        className="side-menu-title"
      >
        React + AFrame
      </h1>
      <hr />
      <div
        className="side-menu-btn-close"
        onClick={[Function]}
      >
        <img
          className="vertical-center"
          src="close-btn.svg"
        />
      </div>
      <div
        className="items-container"
      >
        <div
          className="side-menu-item active"
        >
          <img
            className="icon-item"
            src="start.svg"
          />
          <a
            href="#/"
          >
            Home
          </a>
        </div>
        <div
          className="side-menu-item"
        >
          <img
            className="icon-item"
            src="video-player.svg"
          />
          <a
            href="#/2dvideo"
          >
            2D/3D Video
          </a>
        </div>
        <div
          className="side-menu-item"
        >
          <img
            className="icon-item"
            src="360-degrees.svg"
          />
          <a
            href="#/360video"
          >
            360 Video
          </a>
        </div>
        <div
          className="side-menu-item"
        >
          <img
            className="icon-item"
            src="3d-printer.svg"
          />
          <a
            href="#/3dmodel"
          >
            3D Model Animation
          </a>
        </div>
        <div
          className="side-menu-item"
        >
          <img
            className="icon-item"
            src="help.svg"
          />
          <a
            href="#"
            onClick={[Function]}
          >
            Help
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="top-menu"
  >
    <img
      className="top-menu-icon-left"
      onClick={[Function]}
      src="left-arrow.svg"
    />
    <a
      className="top-menu-item rotate-camera"
      data-position="0.17 4.14 2.79"
    >
      Position 1
    </a>
    <a
      className="top-menu-item rotate-camera"
      data-position="3.48 0.57 0.15"
    >
      Position 2
    </a>
    <a
      className="top-menu-item rotate-camera"
      data-position="-2.89 -2.51 3.20"
    >
      Position 3
    </a>
    <img
      className="top-menu-icon-right"
      onClick={[Function]}
      src="burger-icon.svg"
    />
  </div>
  <a-scene
    cursor="rayOrigin: mouse"
    id="scene"
    raycaster="far: 100; objects: [link], [url]; interval: 200"
  >
    <a-assets>
      <img
        id="sky"
        src="img/1.jpg"
      />
      <img
        id="link1"
        src="img/sea.jpg"
      />
      <img
        id="link2"
        src="img/4.jpg"
      />
      <img
        id="link3"
        src="img/7.jpg"
      />
      <img
        id="aframeArena"
        src="img/aframeArena.png"
      />
    </a-assets>
    <a-sky
      rotation="0 -90 0"
      src="#sky"
    />
    <a-entity
      camera="fov: 80; zoom: 1"
      id="camera"
      orbit-controls="autoRotate:true;target:#entityGroup;enableDamping:true;dampingFactor:0.1;rotateSpeed:0.1;autoRotateSpeed:0.15;zoomSpeed:0.5;minDistance:3;maxDistance:100;invertZoom:true"
      position="0 -0.2 5"
    />
    <a-entity
      id="entityGroup"
    >
      <a-plane
        height="6"
        position="0 -1 0"
        rotation="-90 0 0"
        src="#aframeArena"
        width="6"
      />
      <a-link
        href="#"
        image="#link1"
        onClick={[Function]}
        position="-3 1 0"
        title="2D/3D Video"
      />
      <a-link
        href="#"
        image="#link2"
        onClick={[Function]}
        position="0 1 0"
        title="360 Video"
      />
      <a-link
        href="#"
        image="#link3"
        onClick={[Function]}
        position="3 1 0"
        title="3D Model Animation"
      />
    </a-entity>
  </a-scene>
</main>
`;
